src/logic/otp2/controller.ts

```diff
@@ -303,117 +287,93 @@ function combineAndSortFlexibleAndTransitTripPatterns(
-    if (hasFlexibleResultsOnly) {
-        const transitSearchDate = nextSearchDateFromMetadata || searchDate
-
-        const searchWindow = getMinutesBetweenDates(
-            parseISO(
-                arriveBy
-                    ? flexibleTripPattern.expectedEndTime
-                    : flexibleTripPattern.expectedStartTime,
-            ),
-            transitSearchDate,
-            { min: 60 },
-        )
+    const queries = [getTripPatternsQuery(nextSearchParams)]
+    return {
+        queries,
+        metadata,
+        tripPatterns,
+    }
+}
 
-        const nextSearchParams = {
-            ...getTripPatternsParams,
-            searchDate: transitSearchDate,
-            searchWindow,
-        }
+export async function searchTransitUntilMaxRetries(
+    initialSearchDate: Date,
+    searchParams: Otp2GetTripPatternParams,
+    extraHeaders: { [key: string]: string },
+    prevQueries: GraphqlQuery[],
+): Promise<TransitTripPatterns> {
+    const { searchDate, arriveBy } = searchParams
 
-        const [transitResultsBeforeFlexible, beforeFlexibleMetadata] =
-            await getTripPatterns(nextSearchParams)
+    const queries = [...prevQueries, getTripPatternsQuery(searchParams)]
+    const [response, metadata] = await getTripPatterns(searchParams)
 
-        tripPatterns = transitResultsBeforeFlexible.filter(
-            isValidTransitAlternative,
-        )
-        metadata = beforeFlexibleMetadata
-        queries = [...queries, getTripPatternsQuery(nextSearchParams)]
-    }
+    const tripPatterns = response
+        .filter(isValidTransitAlternative)
+        .map(replaceQuay1ForOsloSWithUnknown)
 
-    tripPatterns = [
-        ...(taxiResults?.tripPatterns || []),
-        ...combineAndSortFlexibleAndTransitTripPatterns(
+    if (tripPatterns.length) {
+        return {
             tripPatterns,
-            nextSearchDateFromMetadata,
-            flexibleTripPattern,
-            arriveBy,
-        ),
-    ]
+            metadata,
+            queries,
+        }
+    }
 
-    if (!tripPatterns.length && metadata) {
+    // Metadata will be null if routingErrors is not empty, because searchWindow
+    // cannot be calculated.
+    //
+    // When we have metadata, we limit by the number of queries as each result
+    // may only be for parts of a day. Queries where metadata is missing is
+    // limited by a set number of days, and each query spans a full day, as we
+    // have no information about the next search window to use.
+    if (metadata) {
         if (queries.length > 15) {
+            // We have tried as many times as we should but could not find
+            // any trip patterns. Give up.
             return {
                 tripPatterns: [],
-                metadata: undefined,
-                hasFlexibleTripPattern: false,
+                metadata: undefined, // Returning metadata undefined will stop generation of a new cursor
```

Comment:
How do we feel about this comment? metadata is optional so it could just be removed, but that exact thing is an important part of how cursors work I think.

Reply:
Keep it! 👍